staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/nonblockinggrpcserver.go

```diff
@@ -36,60 +37,67 @@ type grpcServer struct {
 	grpcVerbosity int
 	wg            sync.WaitGroup
 	endpoint      endpoint
+	socketpath    string
 	server        *grpc.Server
 }
 
 type registerService func(s *grpc.Server)
 
-// endpoint defines where to listen for incoming connections.
+// endpoint defines where and how to listen for incoming connections.
 // The listener always gets closed when shutting down.
 //
-// If the listener is not set, a new listener for a Unix domain socket gets
+// If the listen function is not set, a new listener for a Unix domain socket gets
 // created at the path.
-//
-// If the path is non-empty, then the socket will get removed when shutting
-// down, regardless of who created the listener.
 type endpoint struct {
-	path     string
-	listener net.Listener
+	dir, file string
+	listen    func(ctx context.Context, socketpath string) (net.Listener, error)
 }
 
 // startGRPCServer sets up the GRPC server on a Unix domain socket and spawns a goroutine
 // which handles requests for arbitrary services.
-//
-// The context is only used for additional values, cancellation is ignored.
-func startGRPCServer(valueCtx context.Context, grpcVerbosity int, unaryInterceptors []grpc.UnaryServerInterceptor, streamInterceptors []grpc.StreamServerInterceptor, endpoint endpoint, services ...registerService) (*grpcServer, error) {
-	logger := klog.FromContext(valueCtx)
+func startGRPCServer(logger klog.Logger, grpcVerbosity int, unaryInterceptors []grpc.UnaryServerInterceptor, streamInterceptors []grpc.StreamServerInterceptor, endpoint endpoint, services ...registerService) (*grpcServer, error) {
+	ctx := klog.NewContext(context.Background(), logger)
+
 	s := &grpcServer{
 		endpoint:      endpoint,
 		grpcVerbosity: grpcVerbosity,
 	}
 
-	listener := endpoint.listener
-	if listener == nil {
-		// Remove any (probably stale) existing socket.
-		if err := os.Remove(endpoint.path); err != nil && !os.IsNotExist(err) {
-			return nil, fmt.Errorf("remove Unix domain socket: %v", err)
+	socketpath := path.Join(endpoint.dir, endpoint.file)
+	listen := endpoint.listen
+	if listen == nil {
```

Comment:
Should this logic be attached to the `endpoint` type?

Reply:
Yes. I tried it and it made the code a bit more readable.